### PR TITLE
Datasets: Add Polars Integration 

### PR DESCRIPTION
**Pull request by c-peters**
This PR is now ready for review

### CODE REVIEW

**Comment by severo** [comment]
OK. The CI breaks because the brackets are interpreted by Svelte? Not sure what is the correct way of escaping them? `\{` and `\}`maybe?

> A hugging face url can be constructed from the `username` and `dataset` name like this `hf://datasets/{username}/{dataset}/{path_to_file}`.

**Comment by c-peters** [comment]
Moved the section out of `<Tip></Tip>` assuming that was the problem

File: docs/hub/datasets-libraries.md [hunk 1]
```diff
@@ -13,4 +13,5 @@ The table below summarizes the supported libraries and their level of integratio
 | [DuckDB](./datasets-duckdb) | In-process SQL OLAP database management system.                                                                                                      | ✅ | ✅ |
 | [FiftyOne](./datasets-fiftyone) | FiftyOne is a library for curation and visualization of image, video, and 3D data | ✅ | ✅ |
 | [Pandas](./datasets-pandas) | Python data analysis toolkit.                                                                                                                    | ✅ | ✅ |
+| [Polars](./datasets-polars) |  A DataFrame library on top of an of an OLAP query engine.                                                                                                                    | ✅ | ✅ |
```
Comment (albertvillanova) [inline docs/hub/datasets-libraries.md:16]:
```suggestion
| [Polars](./datasets-polars) |  A DataFrame library on top of an OLAP query engine.                                                                                                                    | ✅ | ✅ |
```

File: docs/hub/datasets-polars-operations.md [hunk 2]
```diff
@@ -0,0 +1,105 @@
+# Transforming your dataset
+
+On this page we'll guide you through some of the most common operations used when doing data analysis. This is only a small subsection of what is possible in Polars, for more information please visit the [Documentation](https://docs.pola.rs/).
+
+For the example we will use the [Danbooru](https://huggingface.co/datasets/KBlueLeaf/danbooru2023-metadata-database) dataset. Danbooru is a large-scale anime image dataset with over 5 million images contributed and annotated in detail by an enthusiast community. Image tags cover aspects like characters, scenes, copyrights, artists, etc with an average of 30 tags per image.
```
Comment (lhoestq) [inline docs/hub/datasets-polars-operations.md:5]: maybe you can showcase a more interesting dataset like the recent Common Crawl statistics at https://huggingface.co/datasets/commoncrawl/statistics ?

File: docs/hub/datasets-polars-file-formats.md [hunk 3]
```diff
@@ -0,0 +1,33 @@
+# File formats
+
+Polars supports the following file formats when reading from Hugging Face:
+
+-  [Parquet](https://docs.pola.rs/api/python/stable/reference/api/polars.read_parquet.html)
```
Comment (ritchie46) [inline docs/hub/datasets-polars-file-formats.md:5]: Arrow/ IPC as well?

File: docs/hub/datasets-polars-file-formats.md [hunk 4]
```diff
@@ -0,0 +1,33 @@
...
+
+Parquet is the preferred file format as it stores the schema with type information within the file. This avoids any ambiguity with parsing and speeds up reading. To read a Parquet file in Polars, use the `read_parquet` function:
+
+```python
+pl.read_parquet("hf://datasets/roneneldan/TinyStories/data/train-00000-of-00004-2d5a1467fff1081b.parquet")
+```
+
+# CSV
+
+The `scan_csv` function can be used to read a CSV file:
```
Comment (ritchie46) [inline docs/hub/datasets-polars-file-formats.md:21]: `scan_csv` should be `read_csv`, given the example.

File: docs/hub/datasets-polars-auth.md [hunk 5]
```diff
@@ -0,0 +1,26 @@
+# Authentication
```
Comment (lhoestq) [inline docs/hub/datasets-polars-auth.md:1]: does Polars support authentication from `huggingface-cli login` ? it stores the token at `~/.cache/huggingface/token` where `~/.cache/huggingface` is the default `HF_HOME`.

File: docs/hub/datasets-polars-file-formats.md [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+pl.read_ndjson("hf://datasets/proj-persona/PersonaHub/persona.jsonl")
+```
+
+# Arrow
+
+Polars supports reading in from an Arrow IPC (Feather v2) file with the `read_ipc` function. 
+
+```python
+pl.read_ipc("hf://...")
+```
```
Comment (lhoestq) [inline docs/hub/datasets-polars-file-formats.md:42]: we don't support IPC format yet (only streaming for legacy reasons), maybe let's remove this for now and we can add back once the Hub supports IPC
```suggestion
```

File: docs/hub/datasets-polars-operations.md [hunk 7]
```diff
@@ -0,0 +1,140 @@
+# Transforming your dataset
```
Comment (lhoestq) [inline docs/hub/datasets-polars-operations.md:1]: love this page !

File: docs/hub/datasets-polars-optimizations.md [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+We will be using the example from the previous page to show the performance benefits of using the lazy API. The code below will compute the number of uploads from `archive.org`.
+
+### Eager
+
+```python
+import polars as pl
+import datetime
+
+df = pl.read_csv("hf://datasets/commoncrawl/statistics/tlds.csv",try_parse_dates = True)
```
Comment (lhoestq) [inline docs/hub/datasets-polars-optimizations.md:19]:
```suggestion
df = pl.read_csv("hf://datasets/commoncrawl/statistics/tlds.csv", try_parse_dates=True)
```

**Review by lhoestq — COMMENTED** [review]
Cool ! Should there be a section on writing to HF as well ? (it can work using `fsspec` I believe)

I added minor suggestions before we merge:

File: docs/hub/datasets-polars-auth.md (hunk 5 above)
Comment (ritchie46) [reply docs/hub/datasets-polars-auth.md:1]: @nameexhaustion do we support this?

Comment (lhoestq) [reply docs/hub/datasets-polars-auth.md:1]: let's keep things as is for now and we can update later if needed

Comment (nameexhaustion) [reply docs/hub/datasets-polars-auth.md:1]: > @nameexhaustion do/can we support this?

made a PR here https://github.com/pola-rs/polars/pull/17881

**Comment by c-peters** [comment]
Putting this as draft, we need to do a Polars python release on our end first

**Comment by c-peters** [comment]
> Putting this as draft, we need to do a Polars python release on our end first

Release is out

**Comment by c-peters** [comment]
> Cool ! Should there be a section on writing to HF as well ? (it can work using `fsspec` I believe)

Writing is not natively supported at the moment, some parts might work with `fsspec`, but I would like to hold off on the section until we add native support

**Review by lhoestq — APPROVED** [review]
Alright let's merge this then ! :)

**Pull request merged**